finality-aleph/src/data_io.rs

```diff
@@ -203,138 +448,535 @@ where
-                    self.add_block(*block_data);
-                    return false;
+            }
+        }
+        false
+    }
+
+    // Returns Err if block is not available.
+    fn get_parent_hash(&mut self, block: &BlockHashNum<B>) -> Result<B::Hash, ()> {
+        if let Some(parent_hash) = self.available_block_cache.get(block) {
+            Ok(*parent_hash)
+        } else if let Some(header) = self
+            .client
+            .header(BlockId::Hash(block.hash))
+            .expect("client must respond")
+        {
+            self.available_block_cache
+                .put(block.clone(), *header.parent_hash());
+            Ok(*header.parent_hash())
+        } else {
+            Err(())
+        }
+    }
+
+    // Returns Err if block is not available.
+    fn get_finalized_at(&mut self, num: NumberFor<B>) -> Result<BlockHashNum<B>, ()> {
+        if self.highest_finalized.num < num {
+            return Err(());
+        }
+        if let Some(hash) = self.finalized_cache.get(&num) {
+            Ok(BlockHashNum::<B>::new(*hash, num))
+        } else if let Some(header) = self
+            .client
+            .header(BlockId::Number(num))
+            .expect("client must respond")
+        {
+            self.finalized_cache.put(num, header.hash());
+            Ok(BlockHashNum::<B>::new(header.hash(), num))
+        } else {
+            Err(())
+        }
+    }
+
+    // Checks whether the parrent of the bottom block of the branch in proposal is finalized.
+    // When calling it, it is assumed that the bottom block is already available.
+    fn is_ancestor_finalized(&mut self, proposal: &AlephProposal<B>) -> bool {
+        let num = proposal.number_below_branch();
+        if self.highest_finalized.num < num {
+            false
+        } else {
+            let bottom = proposal.bottom_block();
+            let parent_hash = self
+                .get_parent_hash(&bottom)
+                .expect("must be available when bottom block is");
+            let finalized = self
+                .get_finalized_at(proposal.number_below_branch())
+                .expect("must be available when bottom block is");
+            parent_hash == finalized.hash
+        }
+    }
+
+    // Outputs the current status of the proposal based on the `old_status` (for optimization).
+    // With `thorough = false` this will only use the cache to check for available/finalized blocks.
+    // With `thorough = true` this will also ask the client for missing blocks.
+    fn check_proposal_availability(
+        &mut self,
+        proposal: &AlephProposal<B>,
+        old_status: &ProposalStatus,
+        thorough: bool,
+    ) -> ProposalStatus {
+        // The implementation of this could possibly be optimized, in particular, always checking
+        // whether it is a hopeless fork is not quite necessary. However, such optimizations could
+        // make the (already complext) code harder to understand and verify correctness.
```

Comment:
Taking into account the other changes happening in this PR this whole proposal is now:

We pass `NonvalidatedProposal`s over the network and validate them using `SessionBounds` into things that are obviously valid (not necessarily correct) to all nodes, regardless of their knowledge about blocks. `AlephProposal` then is type-enforced to be correct – at least by holding hashes in an array, perhaps also by holding the `SessionBounds` and the relative position of the low end of the branch within a session (the latter and variants of it are likely overkill, I'm just trying to get as much type safety as I can; Rust does not have a type system strong enough to make this actually work, so we probably shouldn't try <_<").

Now `AlephProposal` has a method taking `FinalityAndAvailabilityInfo` that returns one of the following:
1. `Finalize(SomethingToFinalize)` – `SomethingToFinalize` can be various things, but likely should be what gets passed to the finalization function. In particular `FinalityAndAvailabilityInfo` should be able to give you that. It might be something slightly weaker if the operation getting you that is expensive.
2. `Pending(WhatWeStillNeed)` – `WhatWeStillNeed` would be a set of requirements that the proposal still needs, like importing or finalizing blocks of specific height or hash.
3. `Ignore` – this would mean the proposal can be let through, but shouldn't be used for finalization (perhaps another name would be better). Mostly used for hopeless forks, but imo also for broken parenthood.
We might also introduce an additional `Incorrect` option that would mean we should discard the data, but I don't think doing this based on block information (rather than just session bounds information) is a good idea – I would rather mark all such things as `Ignore`.

This has the advantage of keeping all the proposal to finalization logic in one place and making it testable (by implementing a mock `FinalityAndAvailabilityInfo`). In data availability we should let through everything that isn't `Pending` and in finalization we should panic (or preferably error and give up within the session) on `Pending` and do the obvious things in the other two cases. This also makes the division into submodules clearer.

Reply:
Ok, I made a lot of changes in the direction of your proposal, but let me first explain why I didn't do exactly what you proposed. This is fairly complex stuff and I spent lots of time thinking about it and implementing it, so what I write below might not be satisfactory, in which case we can schedule a call. But let me try to at least scratch the surface. 

The main problem with what you propose is that we cannot really use the suggested enum output type for both cases of interest (i.e. 1) in `DataStore`, where we check data availability, 2) when interpreting the output of AlephBFT to get blocks to finalize). The reason is that in particular during data availability we could never output `Finalize(Something)` as at that stage there is no way of telling whether some proposal should result in finalizing something or not -- this depends on other proposals ordered. Moreover -- this is perhaps the least intuitive part, but also super-important for correctness of this whole thing: when interpreting the ordered proposals from AlephBFT to get blocks to finalize, **we don't use the client as the source of information about what is finalized and what is not**. Instead we need to keep track of "last finalized block as the result of observing the output of AlephBFT only" (see `last_finalized_by_aleph` in `data_interpreter.rs` in the current version) as the information on `best_finalized` from client could be either:
1) lower than `last_finalized_by_aleph` -- this is expected, and we should make sure we use  `last_finalized_by_aleph`  and not `best_finalized` when interpreting the ordered proposals
2) higher than `last_finalized_by_aleph` -- this could happen if we were behind in finalization and we received justifications of blocks we did not know yet should be finalized by looking at AlephBFT only.
Both cases are tricky, but the implementation what is now implemented by `OrderedDataInterpreter` (pretty much equivalent to the previous version) is I correct.
I think it would be super-messy if we wanted to have a common logic involved for both these things. Or at least I didn't find a good way of doing it.

Instead I created a trait `ProposalStatusProvider` that represents a component for checking the status of a proposal. 

The idea is that you provide the methods:

`fn on_block_imported(&mut self, header: B::Header);`  (Only for optimization)

`fn on_block_finalized(&mut self, block: BlockHashNum<B>);` (Only for optimization)

`fn is_block_imported(&mut self, block: &BlockHashNum<B>) -> bool;`

`fn get_finalized_at(&mut self, number: NumberFor<B>) -> Result<BlockHashNum<B>, ()>;`

`fn get_parent_hash(&mut self, block: &BlockHashNum<B>) -> Result<B::Hash, ()>;`

and the trait then gives you 
`fn get_status(
        &mut self,
        proposal: &AlephProposal<B>,
        old_status: Option<&ProposalStatus>,
    ) -> ProposalStatus `

The main implementation is `CachedPSP` -- this is what `DataStore` did before. There is also a simple implementation that does not do any caching and uses only the client -- `BasicPSP`. 

There are now two components that need a `ProposalStatusProvider`: `DataStore` (data availability) and `OrderedDataInterpreter` (a component that takes ordered data out of AlephBFT and sends blocks to aggregator to create multisigs). 

They are both generic over `ProposalStatusProvider` however I figured that there is no point in actually creating two separate `ProposalStatusProvider`s for them and they simply share a single instance of `CachedPSP` wrapped in a Mutex. 

And btw. I'm using the `parking_lot` sync mutex -- this is a separate discussion (slack), but I think that having the async mutexes here would be much worse.